substrate/client/network/sync/src/engine.rs

```diff
@@ -699,24 +703,24 @@ where
 						removed,
 					)
 				},
-				SyncingAction::CancelBlockRequest { peer_id } => {
-					let removed = self.pending_responses.remove(&peer_id);
+				SyncingAction::CancelRequest { peer_id, key } => {
+					let removed = self.pending_responses.remove(peer_id, key);
```

Comment:
Looking at the code this should not be possible but just to make sure: what can't happen is that there is a ready response waiting which gets canceled and then a new request is sent, replacing the old request. It's not possible for this old, now-canceled response still to be returned to the strategy which it would incorrectly assume it to be a response to the new request?

Reply:
This should not be possible, because the order of actions is preserved: we first cancel the pending response (dropping the response future), and only after initiate a new request.